hello-2.12/hello.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.13.
.TH HELLO "1" "January 2022" "hello 2.12.1-6fe9" "User Commands"
.SH NAME
hello - friendly greeting program
.SH SYNOPSIS
.B hello
[\fI\,OPTION\/\fR]...
.SH DESCRIPTION
Print a friendly, customizable greeting.
.TP
\fB\-t\fR, \fB\-\-traditional\fR
use traditional greeting
.TP
\fB\-g\fR, \fB\-\-greeting\fR=\fI\,TEXT\/\fR
use TEXT as the greeting message
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.SH "ADDITIONAL INFORMATION"
See
.UR https://git.savannah.gnu.org/cgit/hello.git/tree/man/hello.x
man/hello.x
.UE
how to add additional information to a manual page.
.SH AUTHOR
Written by Karl Berry, Sami Kerola, Jim Meyering,
and Reuben Thomas.
.SH "REPORTING BUGS"
Report bugs to: bug\-hello@gnu.org
.br
GNU Hello home page: <https://www.gnu.org/software/hello/>
.br
General help using GNU software: <https://www.gnu.org/gethelp/>
.br
Report GNU Hello translation bugs to <https://translationproject.org/team/>
.SH COPYRIGHT
Copyright \(co 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B hello
is maintained as a Texinfo manual.  If the
.B info
and
.B hello
programs are properly installed at your site, the command
.IP
.B info hello
.PP
should give you access to the complete manual.


hello-2.12/NEWS
GNU Hello NEWS                                         -*- outline -*-

* Noteworthy changes in release 2.12 (2022-01-31) [stable]

Fix a minor documentation typo.

Make tests introduced in 2.11 more portable.

* Noteworthy changes in release 2.11 (2022-01-05) [stable]

Remove -h and -v short options.

Various code tidying, including avoiding forward declarations, removing dead
code, new test for multibyte strings, better use of gnulib, update to more
recent gnulib and add git hooks from coreutils.

* Noteworthy changes in release 2.10 (2014-11-16) [stable]

Most importantly this release include the 'Hello, World' message to be
part of translations.  The translation bug was introduced in release 2.9.

Other changes in this release include; Make use of none-recursive build.
Removal of user-defined new-style.  Include an example how to add a
section to a manual page, such as BUGS.  Rather than 'fprintf (stderr'
use libc 'error()' reporting facility.  Start using 'make
update-copyright' facility.  Generate ChangeLog from git commit logs.
Avoid manual page generation errors when cross-compiling.

* Noteworthy changes in release 2.9 (2013-10-10) [stable]

A series of patches from Sami Kerola has modernized and improved the
build system, added the possibility of user-defined new-style
greetings, and added support for Unicode.

Various translations have been updated (thanks to the Translation Project).

* Noteworthy changes in release 2.8 (2012-04-19) [stable]

The build system has been overhauled to be more in line with
mainstream modern GNU practice, and various translations have been
updated (thanks to the Translation Project!).

Version 2.7 (28 March 2011)
* Distribution:
  . gettext 0.18.1.
  . help2man 1.39.2.
  . updated -translations.

 Version 2.6 (7 April 2010)
* Distribution:
  . make distribution of .tar.xz as well as .tar.gz.
  . gettext 0.17.
  . gnulib updates.
  . use gnulib's srclist-update to keep other common files in sync.
  . updated translations.

 Version 2.5 (7 February 2010)
* Distribution:
  . gnulib updates.
  . autoconf 2.65, automake 1.11.1.
    (Especially including the make dist security fix.)
  . follow latest recommendations for --help and --version.
  . updated translations.

 Version 2.4 (9 December 2008)
* last greeting option wins; before, the precedence was randomly hardwired.
* FDL version 1.3+ for the documentation.
* Distribution:
  . gnulib updates.
  . automake 1.10.2.
  . Include en@quot and en@boldquot support.
  . Remove obsolete translations de_DE and es_AR.
  . new (old) translations: ms (Malay), ro (Romansh), th (Thai).

 Version 2.3 (29 June 2007)
* GPL version 3.

 Version 2.2 (23 November 2006)
* Remove creeping non-feature of reading mail; simplify code accordingly.
* Many improvements to the tests to make them more useful examples.
* --help and --version exit immediately, as specified by GNU standards.
* Distribution:
  . use gnulib for shared source files, such as texinfo.tex and getopt.c.
  . update for automake 1.10, autoconf 2.61, gettext 0.16.
  . new bg (Bulgarian) translation; many updated translations.

 Version 2.1.2 (20 July 2002)
* Updated translations: es, fr, fi, nb, pt_BR, ru, sk, uk.
* New translations: eu, fa, es_AR, ga, ka, ro, sr, zh_CN.

 Version 2.1.1 (9 June 2002)
* Bugfixes:
  . Display the bug reporting address.

 Version 2.1.0 (22 May 2002)
* Automake support.
* Experimental NLS ("Native Language Support").
* Translations for ca, da, de, el, eo, es, et, fi, fr, gl, he, hr, hu,
  id, it, ja, ko, lv, nb, nl, nn, pl, pt, pt_BR, ru, sk, sl, sv, tr, and uk.
* Test suite.
* Add a contrib directory.
* Manual page, produced with 'help2man'.
* Create proper copyright statement for all output formats.
* New greeting message ('-n', '--next-generation').
* Bugfixes:
  . GNU hello ('-m') does not loop forever if MAIL environment variable is
    not set.

 Version 1.3 (22 May 1993)
* GNU hello now exits immediately and successfully after printing its
  version number and/or usage message under the `--version' (or `-v') and
  `--help' (or `-h') options.

 Version 1.2 (21 May 1993)
* GNU hello can now read your mail with the `-m' (or `--mail') option.

 Version 1.1 (16 September 1992)
* There is a new, improved greeting message.
  You can use the new `--traditional' option to get the old behavior.

* Zork output compatibility has been added.

* There is now a `NEWS' file (this one), giving a history of
  user-visible changes.

 Copyright (C) 1992-2019 Free Software Foundation, Inc.

This file is distributed under the same license as the GNU hello package.


hello-2.12/src/hello.c
/* hello.c -- print a greeting message and exit.

   Copyright 1992-2019 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.  */

#include <config.h>

#include <getopt.h>
#include <stdnoreturn.h>
#include <wchar.h>

#include "system.h"

#include "closeout.h"
#include "configmake.h"
#include "dirname.h"
#include "errno.h"
#include "error.h"
#include "gettext.h"
#include "progname.h"
#include "propername.h"
#include "version-etc.h"
#include "xalloc.h"

/* The official name of this program (e.g., no 'g' prefix).  */
#define PROGRAM_NAME "hello"

#define AUTHORS \
  proper_name ("Karl Berry"), \
  proper_name ("Sami Kerola"), \
  proper_name ("Jim Meyering"), \
  proper_name ("Reuben Thomas")

/* Print help info.  This long message is split into
   several pieces to help translators be able to align different
   blocks and identify the various pieces.  */
static _Noreturn void
print_help (FILE *restrict out)
{
  const char *lc_messages = setlocale (LC_MESSAGES, NULL);
  /* TRANSLATORS: --help output 1 (synopsis)
     no-wrap */
  fprintf (out, _("Usage: %s [OPTION]...\n"), program_name);
  /* TRANSLATORS: --help output 2 (brief description)
     no-wrap */
  fputs (_("Print a friendly, customizable greeting.\n"), out);
  fputs ("\n", out);
  /* TRANSLATORS: --help output 3: options
     no-wrap */
  fputs (_("  -t, --traditional       use traditional greeting\n"), out);
  fputs (_("  -g, --greeting=TEXT     use TEXT as the greeting message\n"), out);
  fputs ("\n", out);
  fputs (_("      --help     display this help and exit\n"), out);
  fputs (_("      --version  output version information and exit\n"), out);
  emit_bug_reporting_address();
  /* Don't output this redundant message for English locales.
     Note we still output for 'C' so that it gets included in the man page.  */
  if (lc_messages && STRNCMP_LIT (lc_messages, "en_"))
    {
      /* TRANSLATORS: Replace LANG_CODE in this URL with your language code
	 <https://translationproject.org/team/LANG_CODE.html> to form one of
	 the URLs at https://translationproject.org/team/.  Otherwise, replace
	 the entire URL with your translation team's email address.  */
      fprintf (out, _("Report %s translation bugs to "
		"<https://translationproject.org/team/>\n"), PACKAGE_NAME);
    }
  exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

static void
parse_options (int argc, char *argv[], const char **greeting_msg)
{
  int optc;
  int lose = 0;
  enum {
    OPT_HELP = CHAR_MAX + 1,
    OPT_VERSION
  };
  static const struct option longopts[] = {
    {"greeting", required_argument, NULL, 'g'},
    {"traditional", no_argument, NULL, 't'},
    {"help", no_argument, NULL, OPT_HELP},
    {"version", no_argument, NULL, OPT_VERSION},
    {NULL, 0, NULL, 0}
  };

  while ((optc = getopt_long (argc, argv, "g:t", longopts, NULL)) != -1)
    switch (optc)
      {
	/* --help and --version exit immediately, per GNU coding standards.  */
      case OPT_VERSION:
	version_etc (stdout, PROGRAM_NAME, PACKAGE_NAME, PACKAGE_VERSION, AUTHORS, (char *) NULL);
	exit (EXIT_SUCCESS);
      case 'g':
	*greeting_msg = optarg;
	break;
      case OPT_HELP:
	print_help (stdout);
      case 't':
	*greeting_msg = _("hello, world");
	break;
      default:
	lose = 1;
	break;
      }

  if (lose || optind < argc)
    {
      /* Print error message and exit.  */
      if (argv[optind])
        error (0, 0, "%s: %s", _("extra operand"), argv[optind]);
      emit_try_help ();
      exit (EXIT_FAILURE);
    }
}

int
main (int argc, char *argv[])
{
  const char *greeting_msg;
  wchar_t *mb_greeting;
  mbstate_t mbstate = { 0, };
  size_t len;

  set_program_name (argv[0]);

  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");

#if ENABLE_NLS
  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  /* Having initialized gettext, get the default message. */
  greeting_msg = _("Hello, world!");

  /* Even exiting has subtleties.  On exit, if any writes failed, change
     the exit status.  The /dev/full device on GNU/Linux can be used for
     testing; for instance, hello >/dev/full should exit unsuccessfully.
     This is implemented in the Gnulib module "closeout".  */
  atexit (close_stdout);

  parse_options(argc, argv, &greeting_msg);

  len = strlen(greeting_msg) + 1;
  mb_greeting = xmalloc(len * sizeof(wchar_t));
  len = mbsrtowcs(mb_greeting, &greeting_msg, len, &mbstate);
  if (len == (size_t)-1)
    error (EXIT_FAILURE, errno, _("conversion to a multibyte string failed"));

  /* Print greeting message and exit. */
  wprintf (L"%ls\n", mb_greeting);
  free(mb_greeting);

  exit (EXIT_SUCCESS);
}
